android/modules/platform/src/java/ti/modules/titanium/platform/PlatformModule.java

```diff
@@ -232,24 +242,134 @@ public String createUUID()
 	}
 
 	@Kroll.method
-	public boolean openURL(String url)
+	public boolean canOpenURL(KrollInvocation invocation, String url)
 	{
-		Log.d(TAG, "Launching viewer for: " + url, Log.DEBUG_MODE);
-		Uri uri = Uri.parse(url);
-		Intent intent = new Intent(Intent.ACTION_VIEW, uri);
-		try {
-			Activity activity = TiApplication.getAppRootOrCurrentActivity();
+		// Validate argument.
+		if ((url == null) || url.isEmpty()) {
+			Log.e(TAG, "Ti.Platform.canOpenURL() was given a null or empty string.");
+			return false;
+		}
 
-			if (activity != null) {
-				activity.startActivity(intent);
-			} else {
-				throw new ActivityNotFoundException("No valid root or current activity found for application instance");
+		// Determine if the system has a registered activity intent-filter for the given URL.
+		Intent intent = createOpenUrlIntentFrom(invocation, url);
+		return canOpen(intent);
+	}
+
+	private boolean canOpen(Intent intent)
+	{
+		// Validate argument.
+		if (intent == null) {
+			return false;
+		}
+
+		// If the intent references a local file, then make sure it exists.
+		Uri uri = intent.getData();
+		String scheme = (uri != null) ? uri.getScheme() : null;
+		if (scheme != null) {
+			if (scheme.equals(ContentResolver.SCHEME_CONTENT)) {
+				// We were given a "content://" URL. Check if its ContentProvider can provide file access.
+				// Note: Will typically throw a "FileNotFoundException" or return null if file doesn't exist.
+				ContentResolver contentResolver = TiApplication.getInstance().getContentResolver();
+				if (contentResolver != null) {
+					// First, check if we're referencing an existing file embedded within a file.
+					// Example: A file under the APK's "assets" or "res" folder.
+					boolean wasFileFound = false;
+					try (AssetFileDescriptor descriptor = contentResolver.openAssetFileDescriptor(uri, "r")) {
+						wasFileFound = (descriptor != null);
+					} catch (Exception ex) {
```

Comment:
Maybe log something only for dev/debug purposes. I am not really a fan of empty `catch` blocks. This applies for a few other similar cases in this file.

Reply:
I think empty catch blocks such as these are okay. Unfortunately, I'm forced to guess if the "content://" URL I was given references an `AssetFileDescriptor`, `ParcelFileDescriptor`, or none of the above and it can be only opened via `openInputStream()` which will be the case for in-memory files or streamed files (such as streamed from the cloud). These APIs normally throw a `FileNotFoundException` if I guess wrong instead of simply returning `null`. I don't want to log unnecessary noise in this case since I'm pretty much guaranteed to guess incorrectly at times. (I do the same in the mime-type helper code where if I don't have a file extension, I "guess" it might be an image and attempt to decode it via the `BitmapFactory` class, but it's not a big deal if I guess wrong.)

This is also why I don't log the exception that might be thrown by `canOpenURL()` since that API is used to test if the URL is valid. But... I do log the exception if you call `openURL()` since at that point it is expected to work... and it didn't.

Anyways, that's my reasoning. I'd prefer that exceptions only be used for exceptional circumstances, but Google uses them in places where I wish they wouldn't.